File: mobile/src/main/java/com/windscribe/mobile/ticket/SendTicketActivity.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.ticket;


import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.ViewTreeObserver;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ScrollView;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.appcompat.widget.AppCompatEditText;
import androidx.fragment.app.Fragment;

import com.windscribe.mobile.R;
import com.windscribe.mobile.base.BaseActivity;
import com.windscribe.mobile.custom_view.ErrorFragment;
import com.windscribe.mobile.custom_view.ProgressFragment;
import com.windscribe.mobile.custom_view.SuccessFragment;
import com.windscribe.mobile.di.ActivityModule;
import com.windscribe.vpn.api.response.QueryType;

import java.util.Objects;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import butterknife.OnItemSelected;

public class SendTicketActivity extends BaseActivity
        implements SendTicketView, TextWatcher, ViewTreeObserver.OnGlobalLayoutListener {

    @BindView(R.id.btn_send_ticket)
    Button btnSendButton;

    @BindView(R.id.tv_current_category)
    TextView currentQueryType;

    @BindView(R.id.email)
    AppCompatEditText emailView;

    @BindView(R.id.scroll_view)
    ScrollView mScrollView;

    @BindView(R.id.message)
    AppCompatEditText messageView;

    @Inject
    SendTicketPresenter presenter;

    @BindView(R.id.spinner_query)
    Spinner queryTypeSpinner;

    @BindView(R.id.subject)
    AppCompatEditText subjectView;

    @BindView(R.id.nav_title)
    TextView tvActivityTitle;

    public static Intent getStartIntent(Context context) {
        return new Intent(context, SendTicketActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setActivityModule(new ActivityModule(this, this)).inject(this);
        setContentLayout(R.layout.activity_send_ticket,true);
        presenter.init();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mScrollView.getViewTreeObserver().addOnGlobalLayoutListener(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        mScrollView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
    }

    @Override
    public void addTextChangeListener() {
        messageView.addTextChangedListener(this);
        emailView.addTextChangedListener(this);
        subjectView.addTextChangedListener(this);
    }

    @Override
    public void afterTextChanged(Editable s) {
        String message = Objects.requireNonNull(messageView.getText()).toString();
        String email = Objects.requireNonNull(emailView.getText()).toString();
        String subject = Objects.requireNonNull(subjectView.getText()).toString();
        presenter.onInputChanged(email, subject, message);
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

    }

    @SuppressLint("NonConstantResourceId")
    @OnClick(R.id.nav_button)
    public void onBackButtonPressed() {
        super.onBackPressed();
    }

    @OnClick({R.id.tv_current_category, R.id.img_category_drop_down_btn})
    public void onCurrentQueryTypeClick() {
        queryTypeSpinner.performClick();
    }

    @Override
    public void onGlobalLayout() {
        Rect r = new Rect();
        mScrollView.getWindowVisibleDisplayFrame(r);
        int heightDiff = mScrollView.getRootView().getHeight() - (r.bottom - r.top);
        if (heightDiff > 150) {
            if (mScrollView.getVerticalScrollbarPosition() != mScrollView.getBottom()) {
                mScrollView.post(() -> mScrollView.smoothScrollTo(0, mScrollView.getBottom()));
            }
        }
    }

    @SuppressWarnings("unused")
    @OnItemSelected(R.id.spinner_query)
    public void onQueryTypeSelected(View view, int position) {
        String queryType = queryTypeSpinner.getSelectedItem().toString();
        currentQueryType.setText(queryType);
        presenter.onQueryTypeSelected(QueryType.valueOf(queryType));
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {

    }

    @SuppressLint("NonConstantResourceId")
    @OnClick(R.id.btn_send_ticket)
    public void sendTicketClicked() {
        hideKeyBoard();
        messageView.clearFocus();
        presenter.onSendTicketClicked(Objects.requireNonNull(emailView.getText()).toString(), Objects
                        .requireNonNull(subjectView.getText()).toString(),
                Objects.requireNonNull(messageView.getText()).toString());
    }

    @Override
    public void setActivityTitle(String title) {
        tvActivityTitle.setText(title);
    }

    @Override
    public void setEmail(String email) {
        emailView.setText(email);
    }

    @Override
    public void setErrorLayout(String message) {
        ErrorFragment.getInstance().add(message, this, R.id.cl_overlay, true);
    }

    @Override
    public void setProgressView(boolean show) {
        runOnUiThread(() -> {
            if (show) {
                ProgressFragment.getInstance().add(SendTicketActivity.this, R.id.cl_settings_ticket, true);
            } else {
                Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.cl_settings_ticket);
                if (fragment instanceof ProgressFragment) {
                    ((ProgressFragment) fragment).finishProgress();
                }
            }
        });
    }

    @Override
    public void setQueryTypeSpinner() {
        ArrayAdapter<QueryType> queryAdapter = new ArrayAdapter<>(this,
                R.layout.drop_down_layout,
                R.id.tv_drop_down, QueryType.values());
        queryTypeSpinner.setAdapter(queryAdapter);
    }

    @Override
    public void setSendButtonState(boolean enabled) {
        btnSendButton.setEnabled(enabled);
    }

    @Override
    public void setSuccessLayout(String message) {
        SuccessFragment.getInstance().add(message, this, R.id.cl_overlay, false);
    }

    private void hideKeyBoard() {
        try {
            InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
        } catch (Exception ignored) {
        }
    }
}

File: mobile/src/main/java/com/windscribe/mobile/ticket/SendTicketActivity.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.mobile.ticket

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.appcompat.widget.AppCompatEditText
import butterknife.BindView
import butterknife.OnClick
import butterknife.OnItemSelected
import com.windscribe.mobile.R
import com.windscribe.mobile.base.BaseActivity
import com.windscribe.mobile.di.ActivityModule
import com.windscribe.mobile.dialogs.ErrorDialog
import com.windscribe.mobile.dialogs.ProgressDialog
import com.windscribe.mobile.dialogs.SuccessDialog
import com.windscribe.mobile.welcome.SoftInputAssist
import com.windscribe.vpn.api.response.QueryType
import java.util.*
import javax.inject.Inject

class SendTicketActivity : BaseActivity(), SendTicketView, TextWatcher {
    @Inject
    lateinit var presenter: SendTicketPresenter

    @BindView(R.id.btn_send_ticket)
    lateinit var btnSendButton: Button

    @BindView(R.id.tv_current_category)
    lateinit var currentQueryType: TextView

    @BindView(R.id.email)
    lateinit var emailView: AppCompatEditText

    @BindView(R.id.scroll_view)
    lateinit var scrollView: ScrollView

    @BindView(R.id.message)
    lateinit var messageView: AppCompatEditText

    @BindView(R.id.spinner_query)
    lateinit var queryTypeSpinner: Spinner

    @BindView(R.id.subject)
    lateinit var subjectView: AppCompatEditText

    @BindView(R.id.nav_title)
    lateinit var tvActivityTitle: TextView

    private var softInputAssist: SoftInputAssist? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setActivityModule(ActivityModule(this, this)).inject(this)
        setContentLayout(R.layout.activity_send_ticket, true)
        softInputAssist = SoftInputAssist(this, intArrayOf())
        presenter.init()
    }

    override fun onResume() {
        super.onResume()
        softInputAssist?.onResume()
    }

    override fun onPause() {
        super.onPause()
        softInputAssist?.onPause()
    }

    override fun onDestroy() {
        softInputAssist?.onDestroy()
        super.onDestroy()
    }

    override fun addTextChangeListener() {
        messageView.addTextChangedListener(this)
        emailView.addTextChangedListener(this)
        subjectView.addTextChangedListener(this)
    }

    override fun afterTextChanged(s: Editable) {
        val message = Objects.requireNonNull(
            messageView.text
        ).toString()
        val email = Objects.requireNonNull(
            emailView.text
        ).toString()
        val subject = Objects.requireNonNull(
            subjectView.text
        ).toString()
        presenter.onInputChanged(email, subject, message)
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

    @SuppressLint("NonConstantResourceId")
    @OnClick(R.id.nav_button)
    fun onBackButtonPressed() {
        super.onBackPressed()
    }

    override fun onBackPressed() {
        setInputState(true)
        super.onBackPressed()
    }

    @OnClick(R.id.tv_current_category, R.id.img_category_drop_down_btn)
    fun onCurrentQueryTypeClick() {
        queryTypeSpinner.performClick()
    }

    @OnItemSelected(R.id.spinner_query)
    fun onQueryTypeSelected() {
        val queryType = queryTypeSpinner.selectedItem.toString()
        currentQueryType.text = queryType
        QueryType.values()[queryTypeSpinner.selectedItemPosition].let {
            presenter.onQueryTypeSelected(it)
        }
    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

    @SuppressLint("NonConstantResourceId")
    @OnClick(R.id.btn_send_ticket)
    fun sendTicketClicked() {
        hideKeyBoard()
        messageView.clearFocus()
        presenter.onSendTicketClicked(
            Objects.requireNonNull(
                emailView.text
            ).toString(), Objects
                .requireNonNull(subjectView.text).toString(),
            Objects.requireNonNull(messageView.text).toString()
        )
    }

    override fun setActivityTitle(title: String) {
        tvActivityTitle.text = title
    }

    override fun setEmail(email: String) {
        emailView.setText(email)
    }

    private fun setInputState(isEnabled: Boolean) {
        queryTypeSpinner.isEnabled = isEnabled
        emailView.isEnabled = isEnabled
        subjectView.isEnabled = isEnabled
        btnSendButton.isEnabled = isEnabled
        messageView.isEnabled = isEnabled
    }

    override fun setErrorLayout(message: String) {
        setInputState(false)
        ErrorDialog.show(this, message)
    }

    override fun setProgressView(show: Boolean) {
        if (show) {
            ProgressDialog.show(this)
        } else {
            ProgressDialog.hide(this)
        }
    }

    override fun setQueryTypeSpinner() {
        val queryAdapter = ArrayAdapter(
            this,
            R.layout.drop_down_layout,
            R.id.tv_drop_down,
            resources.getStringArray(R.array.query_types)
        )
        queryTypeSpinner.adapter = queryAdapter
    }

    override fun setSendButtonState(enabled: Boolean) {
        btnSendButton.isEnabled = enabled
    }

    override fun setSuccessLayout(message: String) {
        setInputState(false)
        SuccessDialog.show(this, message, null)
    }

    private fun hideKeyBoard() {
        try {
            currentFocus?.windowToken?.let {
                val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(it, 0)
            }
        } catch (ignored: Exception) {
        }
    }

    companion object {
        fun getStartIntent(context: Context?): Intent {
            return Intent(context, SendTicketActivity::class.java)
        }
    }
}